File: return_calculator.py
from dateutil.parser import parse

class ReturnCalculator:
  def __init__(self, filename):
    self.snapshots = self.read_snapshots(filename)

  def calculate(self):
    snapshots = self.snapshots
    if (len(snapshots) < 2):
      return 0
    runningProduct = 1
    for i in range(len(snapshots) - 1):
      if (snapshots[i]['market_value'] == 0):
        return 0
      runningProduct *= (snapshots[i+1]['market_value'] - snapshots[i+1]['cash_flow']) / snapshots[i]['market_value']
    return runningProduct - 1

  def read_snapshots(self, filename):
    fil = open(filename, 'r')
    snapshots = []
    for line in fil:
      row = line.split(',')
      date = parse(row[0])
      cash_flow = float(row[1])
      market_value = float(row[2])
      snapshots.append({ 'date': date, 'cash_flow': cash_flow, 'market_value': market_value })
    return snapshots

a = ReturnCalculator('test/test-0.csv')
print a.calculate()

File: ReturnCalculator.java
import java.util.*;
import java.io.*;

public class ReturnCalculator {
    public static final String fileName = "test/test-0.csv";

    public static void main(String[] args) throws FileNotFoundException {
        Scanner scanner = new Scanner(new File(fileName));

        List<Snapshot> snapshots = new ArrayList<>();
        while (scanner.hasNext()) {
            snapshots.add(new Snapshot(scanner.nextLine()));
        }

        List<Double> subperiodNaiveReturns = new ArrayList<>();
        for (int i = 0, j = 1; j < snapshots.size(); i++, j++) {
            Snapshot day1 = snapshots.get(i);
            Snapshot day2 = snapshots.get(j);
            subperiodNaiveReturns.add(getNaiveReturn(day1, day2));
        }

        double compoundedReturn = 1;
        for (double subperiodNaiveReturn : subperiodNaiveReturns) {
            compoundedReturn *= (subperiodNaiveReturn >= 0) 
                ? (1 + subperiodNaiveReturn) 
                : (1 - subperiodNaiveReturn);
        }
        System.out.println(compoundedReturn);
    }

    private static double getNaiveReturn(Snapshot day1, Snapshot day2) {
        return ((day2.marketValue - day2.cashFlow) / day1.marketValue) - 1;
    }

    private static class Snapshot {
        public String date;
        public double cashFlow;
        public double marketValue;

        public Snapshot(String string) {
            // Parse string.
            String[] separated = string.split(",");

            // TODO: Throw exception if length less than 3.

            this.date = separated[0];
            this.cashFlow = Double.parseDouble(separated[1]);
            this.marketValue = Double.parseDouble(separated[2]);
        }
    }
}
